clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1330,6 +1330,27 @@ event_pre_syscall(void *drcontext, int sysnum)
         return true;
     if (BUF_PTR(data->seg_base) == NULL)
         return true; /* This thread was filtered out. */
+
+    // If we just switched modes in this thread in a block ending in a syscall,
+    // we can end up here without having traced the syscall instr, which we
+    // want to avoid.  We look for an empty (non-initial, so we do not add
+    // init_header_size) buffer with zero instrs in the current window (b/c it
+    // could be empty just due to the syscall instr filling up the prior buffer).
+    // (The converse can happen, with a tracing window ending in a syscall but
+    // the mode having changed, causing us to exit up above and not emit a marker:
+    // we solve that by removing syscalls without markers in raw2trace.)
+    if (BUF_PTR(data->seg_base) == data->buf_base + buf_hdr_slots_size &&
```

Comment:
Make this a separate function (is_buffer_window_empty)

Reply:
Done.